### DISCUSSION TITLE
What is the structure of a : dictionary entry?  Is it a typical Forth structure or some hybrid?

### DISCUSSION THREAD

**Discussion by marklacas** [tabemann/zeptoforth#28] [Q&A]
I have yet to find in any documentation for zeptoForth a diagram of a typical : dictionary entry.

**Comment by tabemann** [comment]
zeptoforth has two conventional Forth-type linked list dictionaries, one for words in RAM and one for words in flash.

On the RP2040 it also has a "minidictionary" in RAM for words in flash to get around the limitations of the external flash. It consists of a hash table where each entry consists of two cells, the first a hash of the word's name and wordlist ID and the second a pointer to the actual word header in flash.

The structure of the primary word headers can be found in `src/common/forth/basic.fs`. It is as follows:

```
\ Get the address of the flags for a word
: word-flags ( word -- h-addr ) [inlined] ;

\ Get the address of the wordlist id for a word
: wordlist-id ( word -- h-addr ) 2 + [inlined] ;

\ Get the next word for a word
: next-word ( word -- addr ) 4 + [inlined] ;

\ Get the name of a word (a counted string)
: word-name ( word -- c-addr ) 8 + [inlined] ;
```

Note that all these words are in the `internal` module. `word-name` points to a counted string, which is followed by a halfword-aligned executable word body.

The flags are also defined in `src/common/forth/basic.fs`. They are:

```
\ Visible flag
1 constant visible-flag

\ Immediate flag
2 constant immediate-flag

\ Compile-only flag
4 constant compiled-flag

\ Inlined flag
8 constant inlined-flag

\ Folded flag
16 constant fold-flag

\ Initialized value flag
32 constant init-value-flag
```

**Comment by marklacas** [comment]
Thanks!
